### ISSUE TITLE
Renderer added generated names to nodes and relations inside conditions which lead to InvalidDataAccessResourceUsageException.

### ISSUE THREAD

**Issue by gexterio** [neo4j-contrib/cypher-dsl#1014]
I use kotlin + Spring Boot 3.3.0 with org.neo4j:neo4j-cypher-dsl:2023.9.5.

```kotlin
class BookRepositoryCypherDslQueryProvider {

    private val userIdParam = parameter("userId")
    private val limitParam = parameter("limit")
    private val langParam = parameter("lang")


    private val book: Book_ = BOOK.withProperties(BOOK.LANG, langParam)
        .named("book")
    private val userDetails: UserDetails_ =
        USER_DETAILS.withProperties(USER_DETAILS.ID, userIdParam)
            .named("user")
    private val userPreferences: UserPreferences_ =
        USER_PREFERENCES.withProperties(USER_PREFERENCES.USER_ID, userIdParam)
            .named("preferences")
    private val genre: Genre_ = GENRE.named("genre")
    private val userActivity: UserSuggestionActivity_ = USER_SUGGESTION_ACTIVITY.withProperties(
        USER_SUGGESTION_ACTIVITY.USER_ID, userIdParam
    ).named("activity")

    private val relationBookGenre: RelationshipPattern = book.relationshipTo(genre, "BELONGS_TO").named("rel")

    private val relationBookAvoidedGenre:RelationshipPattern =
        book.relationshipTo(GENRE, "BELONGS_TO").relationshipFrom(userPreferences, "AVOIDED")

    private val relationUserPreferencesGenreBook:RelationshipPattern = userDetails
        .relationshipTo(userPreferences, "HAS_PREFERENCES")
        .relationshipTo(genre, "PREFERRED")
        .relationshipFrom(book, "BELONGS_TO")

    private val returningBooks:Expression = call("distinct").withArgs(name("book")).asFunction()


    fun buildStatement(
        userId: UUID,
        lang: String,
        limit: Int,
        includeAvoided: Boolean,
        includePreferred: Boolean,
    ): Statement {
        val statement = match(
            relationUserPreferencesGenreBook
        ).where(not(exists(relationBookAvoidedGenre)))
            .with(book, userActivity).limit(limitParam)
            .match(relationBookGenre)
            .where(not(exists(book.relationshipBetween(userActivity))))
            .returning(
                returningBooks, collect(name("rel")), collect(genre)
            )
            .build()

        return statement
    }
}

```
To execute I use `CypherdslStatementExecutor.findAll(statement: Statement)` function.
This statement leads  to exception while executing:
```log
Request processing failed: org.springframework.dao.InvalidDataAccessResourceUsageException: PatternExpressions are not allowed to introduce new variables: 'cdMZWlfQ008'. (line 1, column 278 (offset: 277))
"MATCH (user:`UserDetails` {id: $userId})-[undWpzxO001:`HAS_PREFERENCES`]->(preferences:`UserPreferences` {userId: $userId})-[DHYgMrye003:`PREFERRED`]->(genre:`Genre`)<-[jDhOJrQo005:`BELONGS_TO`]-(book:`Book` {lang: $lang}) WHERE NOT (exists((book)-[KUXCWZni007:`BELONGS_TO`]->(cdMZWlfQ008)<-[FhHrdBZj009:`AVOIDED`]-(preferences))) WITH book, activity LIMIT $limit MATCH (book)-[rel:`BELONGS_TO`]->(genre:`Genre`) WHERE NOT (exists((book)--(activity))) RETURN distinct(book), collect(rel), collect(genre)"
```

In manually, for example using neo4j Browser i could use unnamed nodes/relations to describe condition and it will work well. 
Is it possible to avoid setting names here?

**Comment by michael-simons** [comment]
Thanks for your report. 

This is actually a bug in which a name is generated for a node used in a pattern *expression*. This needs a fix.

The fix won't be turning off name generation, it is needed in several SDN specific use cases.

**Comment by gexterio** [comment]
It's also may be my mistaken usage. If i add `.asCondition()` to the end of defined pattern, names are not generating.

**Comment by michael-simons** [comment]
Hm, @gexterio would you please share your domain classes? I have an idea, but I cannot fully reproduce it. Reason is in how the static meta model is generated.

Btw Your query will fail nevertheless, as you return the activity but never matched it.

Thanks.

**Comment by michael-simons** [comment]
(FWIW I can reproduce in part, but not on the `BELONGS_TO` relationship, and that is the only one you didn't create using the static meta model in the partial code above), the output you show doesn't match the code for me

**Comment by gexterio** [comment]
@michael-simons , no problem. And about your second message: Sorry for mistaken. I fixed this in the code but forgot to update here.

```kotlin
@Node
data class Book(

    @Id @GeneratedValue var id: UUID? = null,

    var title: String,

    var description: String,

    var imageUrl: String,

    var ageLimit: Int,

    var publishedAt: Int,

    var storeUrl: String,

    var isbn: String,

    var lang: String,

    @Relationship(type = "BELONGS_TO", direction = Relationship.Direction.OUTGOING)
    var genres: Set<Genre>,

    var author: String
)

@Node
data class UserDetails(

    @Id val id: UUID,

    val userName: String,

    val email: String,

    val firstName: String? = null,

    val lastName: String? = null,

    val dateOfBirth: LocalDate? = null,

    val gender: Gender? = null,
)

@Node
data class Genre(
    @Id var name: String
)

@Node
class UserPreferences(
    @Id var userId: UUID? = null,

    @Relationship(type = "HAS_PREFERENCES", direction = Relationship.Direction.INCOMING)
    var userDetails: UserDetails? = null,

    @Relationship(type = "PREFERRED", direction = Relationship.Direction.OUTGOING)
    var preferredGenres: Set<Genre>? = emptySet(),

    @Relationship(type = "AVOIDED", direction = Relationship.Direction.OUTGOING)
    var avoidedGenres: Set<Genre>? = emptySet()
)

@Node
data class UserSuggestionActivity(

    @Id var userId: UUID? = null,

    @Relationship(type = "HAS_ACTIVITY", direction = Relationship.Direction.INCOMING)
    var userDetails: UserDetails? = null,

    @Relationship(type = "LIKED", direction = Relationship.Direction.OUTGOING)
    var likedBooks: MutableSet<Book>? = mutableSetOf(),

    @Relationship(type = "DISLIKED", direction = Relationship.Direction.OUTGOING)
    var dislikedBooks: MutableSet<Book>? = mutableSetOf(),

    @Relationship(type = "SAVED", direction = Relationship.Direction.OUTGOING)
    var savedBooks: MutableSet<Book>? = mutableSetOf(),

    @Relationship(type = "SKIPPED", direction = Relationship.Direction.OUTGOING)
    var skippedBooks: MutableSet<Book>? = mutableSetOf()
)
```

**Comment by michael-simons** [comment]
> Sorry for mistaken. I fixed this in the code but forgot to update here.

What then is the final version that creates the issue? 

Also thanks for the update above

**Comment by gexterio** [comment]
I'ms sorry for the huge code. For now i solved this problem, i suppose just by adding `asCondition`.

```kotlin
@Component
class BookRepositoryCypherDslQueryProvider {

    private val userIdParam = parameter("userId")
    private val limitParam = parameter("limit")
    private val langParam = parameter("lang")


    private val book: Book_ = BOOK.withProperties(BOOK.LANG, langParam)
        .named("book")
    private val userDetails: UserDetails_ =
        USER_DETAILS.withProperties(USER_DETAILS.ID, userIdParam)
            .named("user")
    private val userPreferences: UserPreferences_ =
        USER_PREFERENCES.withProperties(USER_PREFERENCES.USER_ID, userIdParam)
            .named("preferences")
    private val genre: Genre_ = GENRE.named("genre")
    private val userActivity: UserSuggestionActivity_ = USER_SUGGESTION_ACTIVITY.withProperties(
        USER_SUGGESTION_ACTIVITY.USER_ID, userIdParam
    ).named("activity")

    private val relationBookGenre: RelationshipPattern = book.relationshipTo(genre, "BELONGS_TO").named("rel")

    private val relationBookAvoidedGenre =
        book.relationshipTo(genre, "BELONGS_TO").relationshipFrom(userPreferences, "AVOIDED").asCondition().asCondition()

    private val relationUserPreferencesGenreBook:RelationshipPattern = userDetails
        .relationshipTo(userPreferences, "HAS_PREFERENCES")
        .relationshipTo(genre, "PREFERRED")
        .relationshipFrom(book, "BELONGS_TO")

    private val returningBooks:Expression = call("distinct").withArgs(name("book")).asFunction()


    fun buildStatement(
        userId: UUID,
        lang: String,
        limit: Int,
        includeAvoided: Boolean,
        includePreferred: Boolean,
    ): Statement {
        val statement = match(
            relationUserPreferencesGenreBook
        ).match(userDetails.relationshipTo(userActivity, "HAS_ACTIVITY"))
            .where(not(exists(relationBookAvoidedGenre).asCondition())) //here I add `asCodnition`
            .with(book, userActivity).limit(limitParam)
            .match(relationBookGenre)
            .where(not(exists(book.relationshipBetween(userActivity))))
            .returning(
                returningBooks, collect(name("rel")), collect(genre)
            )
            .build()

        return statement
    }
}
```

**Comment by michael-simons** [comment]
Thanks again, and no need to apologise, your input was very valuable. 

I have PR fix ready for the main branch, which I will backport to 2023.9.x just in case.

**Comment by michael-simons** [comment]
2023.9.8 has been released and will be part of SDN 7.3.x onwards.
Should also be a drop in for 7.2.x.
